src/us/kbase/workspace/database/provenance/Title.java

```diff
@@ -0,0 +1,171 @@
+ */
+public class Title {
+
+	private final String titleString;
+	private final TitleType titleType;
+	private final String titleLanguage;
+
+	/**
+	 * An enum representing types of title.
+	 */
+	public enum TitleType {
+		SUBTITLE,
+		ALTERNATIVE_TITLE,
+		TRANSLATED_TITLE,
+		OTHER;
+
+		private static final Map<String, TitleType> STRING_TO_TITLE_TYPE_MAP = new HashMap<>();
+		static {
+			for (final TitleType tt: TitleType.values()) {
+				STRING_TO_TITLE_TYPE_MAP.put(tt.name().toLowerCase(), tt);
+			}
+		}
+
+		/** Gets a title type based on a supplied string.
+		 * @param input a string representing a title type.
+		 * @return a TitleType.
+		 * @throws IllegalArgumentException if there is no title type related to the input string.
+		 */
+		public static TitleType getTitleType(final String input) {
+			final String lowercaseInput = Util.checkString(input, "input").toLowerCase();
+			if (STRING_TO_TITLE_TYPE_MAP.containsKey(lowercaseInput)) {
+				return STRING_TO_TITLE_TYPE_MAP.get(lowercaseInput);
+			}
+			throw new IllegalArgumentException("Invalid titleType: " + input);
+		}
+	}
+
+	private Title(
+			final String titleString,
+			final TitleType titleType,
+			final String titleLanguage) {
+		this.titleString = titleString;
+		this.titleLanguage = titleLanguage;
+		this.titleType = titleType;
+	}
+
+	/**
+	 * Gets the title of the resource.
+	 *
+	 * @return the titleString.
+	 */
+	public String getTitleString() {
+		return titleString;
+	}
+
+	/**
+	 * Gets the type of title being represented.
+	 *
+	 * @return the titleType, if present.
+	 */
+	public Optional<TitleType> getTitleType() {
+		return Optional.ofNullable(titleType);
+	}
+
+	/**
+	 * Gets the language in which the title is written.
+	 *
+	 * @return the titleLanguage, if present.
+	 */
+	public Optional<String> getTitleLanguage() {
+		return Optional.ofNullable(titleLanguage);
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(titleLanguage, titleString, titleType);
+	}
+
+	@Override
+	public boolean equals(Object obj) {
+		if (this == obj)
+			return true;
+		if (obj == null)
+			return false;
+		if (getClass() != obj.getClass())
+			return false;
+		Title other = (Title) obj;
+		return Objects.equals(titleLanguage, other.titleLanguage)
+				&& Objects.equals(titleString, other.titleString)
+				&& Objects.equals(titleType, other.titleType);
+	}
+
+	/**
+	 * Gets a builder for an {@link Title}.
+	 *
+	 * @return the builder.
+	 */
+	public static Builder getBuilder(final String titleString) {
+		return new Builder(titleString);
+	}
+
+	/** A builder for an {@link Title}. */
+	public static class Builder {
+
+		private String titleString;
+		private TitleType titleType = null;
+		private String titleLanguage = null;
+
+		private Builder(final String titleString) {
+			this.titleString = Util.checkString(titleString, "titleString");
+		}
+
+		/**
+		 * Sets the language of title being represented, for example en-GB.
```

Comment:
if you're using IETF format for language identifiers it'd be great to check that at least

Reply:
I plan to retrofit the language codes as an enum -- I haven't yet found a good source but it's on the to-do list.

Reply:
https://stackoverflow.com/questions/60795579/validating-bcp-47-language-tag-using-java-apis didn't read closely but might do what you want

Reply:
Also https://docs.oracle.com/javase/8/docs/api/java/util/Locale.html

Reply:
https://docs.oracle.com/javase/8/docs/api/java/util/Locale.html#forLanguageTag-java.lang.String-

Reply:
Great, thanks for those links!

Reply:
After much research: Java does not have a BCP-47 validator; I can check that the input has the appropriate syntax but there's no built-in method for asserting that the values are valid language/region/script/extlang tags. I would need to pull in the [source files](https://www.iana.org/assignments/language-subtag-registry/language-subtag-registry) and build my own parser/validator. If I'm going to do that, I think I'd prefer to either use an existing lib (there are some py libs available) or parse/convert the file to something like JSONschema, which can be used elsewhere, not just in these Java classes.

Reply:
`Locale` doesn't do what you want?

Reply:
No, it just checks that the syntax is OK. I've used its capabilities such as they are, so at least the values can be checked for conformance with BCP-47 syntax rules.

Reply:
Well, that's poopy